centreon-certified/canopsis/canopsis2x-events-apiv2.lua

```diff
@@ -0,0 +1,868 @@
+    :setopt_writefunction(
+      function (response)
+        http_response_body = http_response_body .. tostring(response)
+      end
+    )
+    :setopt(curl.OPT_TIMEOUT, queue.sc_params.params.connection_timeout)
+    :setopt(curl.OPT_SSL_VERIFYPEER, queue.sc_params.params.allow_insecure_connection)
+    :setopt(curl.OPT_HTTPHEADER, queue_metadata.headers)
+    :setopt(curl.OPT_CUSTOMREQUEST, "GET")
+
+  -- set proxy address configuration
+  if (queue.sc_params.params.proxy_address ~= '') then
+    if (queue.sc_params.params.proxy_port ~= '') then
+      http_request:setopt(curl.OPT_PROXY, queue.sc_params.params.proxy_address .. ':' .. queue.sc_params.params.proxy_port)
+    else
+      queue.sc_logger:error("[getCanopsisAPI]: proxy_port parameter is not set but proxy_address is used")
+    end
+  end
+
+  -- set proxy user configuration
+  if (queue.sc_params.params.proxy_username ~= '') then
+    if (queue.sc_params.params.proxy_password ~= '') then
+      http_request:setopt(curl.OPT_PROXYUSERPWD, queue.sc_params.params.proxy_username
+        .. ':' .. queue.sc_params.params.proxy_password)
+    else
+      queue.sc_logger:error("[getCanopsisAPI]: proxy_password parameter is not set but proxy_username is used")
+    end
+  end
+
+  -- performing the HTTP request
+  http_request:perform()
+
+  -- collecting results
+  http_response_code = http_request:getinfo(curl.INFO_RESPONSE_CODE)
+
+  http_request:close()
+
+  if http_response_code == 200 then
+    queue.sc_logger:info("[getCanopsisAPI]: HTTP request successful: return code is "
+      .. tostring(http_response_code))
+    -- now handling response content which should be JSON
+    local json_response_decoded, error = broker.json_decode(http_response_body)
+    if error then
+      queue.sc_logger:error("[getCanopsisAPI]: couldn't decode json string: " .. tostring(http_response_body)
+        .. ". Error is: " .. tostring(error))
+      return retval
+    end
+    -- Handle Type
+    if type_name ~= "" and reason_name == "" then
+      for json_element, type_object in pairs(json_response_decoded["data"]) do
+        if type_object["name"] == type_name then
+          retval = type_object["_id"]
+        end
+      end
+    -- Handle Reason
+    elseif type_name == "" and reason_name ~= "" then
+      for json_element, reason_object in pairs(json_response_decoded["data"]) do
+        if reason_object["name"] == reason_name then
+          retval = reason_object["_id"]
+        end
+      end
+    end
+  else
+    queue.sc_logger:error("[getCanopsisAPI]: HTTP request FAILED, return code is "
+      .. tostring(http_response_code) .. ". Message is: " .. tostring(http_response_body))
+  end
+
+  return retval
+end
+
+--------------------------------------------------------------------------------
+-- Required functions for Broker StreamConnector
+--------------------------------------------------------------------------------
+
+local queue
+
+-- Fonction init()
+function init(conf)
```

Comment:
this part is not that important, maybe for the sake of standardization it kind be important.

I like to have the init(conf) function, kind of empty. It bothers me to use the queue name for the object. Every time I'm using self and forget that in those global fucntions, I should use queue.

So, all the the work that you are doing here, I'd like to put it inside the EventQueue:enw function

I had a similar case where I wanted to do a lot of checks/pre-configuration during the init phase. This was for the elastic metrics stream connector. 

Quick explaination here: 
I had to check if indexes were created in elasticsearch and if not create them if a parameter was asking me to do it
https://github.com/centreon/centreon-stream-connector-scripts/blob/079d2706d25573bd4caf69e6dc76350809bb328a/centreon-certified/elasticsearch/elastic-metrics-apiv2.lua#L140

and then, If for any reason it failed, I put the fail flag with a log error. (that something that you should add in your mecanism)

I have just added a mecanism to avoid spam in the log (We should probably find a better way to do it, It was a quick one shot and should improved and set in every stream connector later on)

https://github.com/centreon/centreon-stream-connector-scripts/blob/079d2706d25573bd4caf69e6dc76350809bb328a/centreon-certified/elasticsearch/elastic-metrics-apiv2.lua#L714

Reply:
Done [here](https://github.com/centreon/centreon-stream-connector-scripts/pull/194/commits/b158b4e22b057aad1b11365f96b7e098e0eea98b)